syntax = "proto3";

package grpc_pricer;

// Définition du service
service GrpcPricer {
	rpc PriceAndDeltas (PricingInput) returns (PricingOutput);
	rpc HelloWorld (Empty) returns (ReqInfo);
}

// Message contenant les données pour invoquer le pricer
message PricingInput {

  	repeated PastLines past = 1;  // Matrice des données historiques
  	bool monitoringDateReached = 2; // Indique si une date de monitoring est atteinte
  	double time = 3; // Temps mathématique
	repeated Currency currencies = 4;
	string domesticCurrencyId = 5;
	repeated Asset assets = 6;
	int32 numberOfDaysInOneYear = 7; //  252 par déafut :
	PortfolioRebalancingOracleDescription portfolioRebalancingOracleDescription = 8;
	Option option = 9; // ? :
	repeated CorrelationMatrix correlations = 10;
	// int32 sampleNb = 11; 50000
	// double relativeFiniteDifferenceStep = 12; // 0.1 

}

// Une ligne de la matrice historique
message PastLines {
	repeated double value = 1;
}

// Informations sur les devises
message Currency {
	string id = 1;
	double interestRate = 2;
	double volatility = 3;
}

// Informations sur les actifs financiers
message Asset {
	string currencyId = 1;
	double volatility = 2;
}

// Description de la stratégie de rééquilibrage du portefeuille
message PortfolioRebalancingOracleDescription {
	string type = 1;
	int32 period = 2;
}

// Détails de l'option
message Option {
	string type = 1;
	FixingDatesInDays fixingDatesInDays = 2;
	double strike = 3;
	int32 maturityInDays = 4;
}

// Fixing dates en jours
message FixingDatesInDays {
	string type = 1;
	repeated int32 datesInDays = 2;
}

// Matrice des corrélations
message CorrelationMatrix {
	repeated double values = 1; // Stocke une ligne de la matrice des corrélations
}

// Résultats du pricing
message PricingOutput {
	double price = 1;  // Prix calculé
	repeated double deltas = 2;  // Liste des deltas
	double priceStdDev = 3;  // Écart-type du prix
	repeated double deltasStdDev = 4;  // Écart-type des deltas
}

// Message pour le Heartbeat
message Empty {}

// ReqInfo : pour vérifier que Grpc est bien configuré
message ReqInfo {
	string message = 1 ; 
}
